core/embed/rust/src/protobuf/defs.rs

```diff
@@ -0,0 +1,208 @@
+impl FieldType {
+    pub fn primitive_type(&self) -> u8 {
+        match self {
+            FieldType::UVarInt | FieldType::SVarInt | FieldType::Bool | FieldType::Enum(_) => {
+                PRIMITIVE_TYPE_VARINT
+            }
+            FieldType::Bytes | FieldType::String | FieldType::Msg(_) => {
+                PRIMITIVE_TYPE_LENGTH_DELIMITED
+            }
+        }
+    }
+}
+
+pub struct EnumDef {
+    pub values: &'static [u16],
+}
+
+static ENUM_DEFS: &[u8] = include_bytes!("../../../../src/trezor/messages/proto_enums.data");
+static MSG_DEFS: &[u8] = include_bytes!("../../../../src/trezor/messages/proto_msgs.data");
+static NAME_DEFS: &[u8] = include_bytes!("../../../../src/trezor/messages/proto_names.data");
+static WIRE_DEFS: &[u8] = include_bytes!("../../../../src/trezor/messages/proto_wire.data");
+
+fn find_msg_offset_by_name(msg_name: u16) -> Option<u16> {
+    #[repr(C, packed)]
+    struct NameDef {
+        msg_name: u16,
+        msg_offset: u16,
+    }
+
+    let name_defs: &[NameDef] = unsafe {
+        slice::from_raw_parts(
+            NAME_DEFS.as_ptr().cast(),
+            NAME_DEFS.len() / mem::size_of::<NameDef>(),
+        )
+    };
+    name_defs
+        .binary_search_by_key(&msg_name, |def| def.msg_name)
+        .map(|i| name_defs[i].msg_offset)
+        .ok()
+}
+
+fn find_msg_offset_by_wire(wire_id: u16) -> Option<u16> {
+    #[repr(C, packed)]
+    struct WireDef {
+        wire_id: u16,
+        msg_offset: u16,
+    }
+
+    let wire_defs: &[WireDef] = unsafe {
+        slice::from_raw_parts(
+            WIRE_DEFS.as_ptr().cast(),
+            WIRE_DEFS.len() / mem::size_of::<WireDef>(),
+        )
+    };
+    wire_defs
+        .binary_search_by_key(&wire_id, |def| def.wire_id)
+        .map(|i| wire_defs[i].msg_offset)
+        .ok()
+}
+
+unsafe fn get_msg(msg_offset: u16) -> MsgDef {
+    // #[repr(C, packed)]
+    // struct MsgDef {
+    //     fields_count: u8,
+    //     defaults_size: u8,
+    //     flags_and_wire_id: u16,
+    //     fields: [Field],
+    //     defaults: [u8],
+    // }
+
+    // SAFETY: `msg_offset` has to point to a beginning of a valid message
+    // definition inside `MSG_DEFS`.
+    unsafe {
+        let ptr = MSG_DEFS.as_ptr().add(msg_offset as usize);
+        let fields_count = ptr.offset(0).read() as usize;
+        let defaults_size = ptr.offset(1).read() as usize;
+
+        let flags_and_wire_id_lo = ptr.offset(2).read();
+        let flags_and_wire_id_hi = ptr.offset(3).read();
+        let flags_and_wire_id = u16::from_le_bytes([flags_and_wire_id_lo, flags_and_wire_id_hi]);
+
+        let is_experimental = flags_and_wire_id & 0x8000 != 0;
+        let wire_id = match flags_and_wire_id & 0x7FFF {
+            0x7FFF => None,
+            some_wire_id => Some(some_wire_id),
+        };
+
+        let fields_size = fields_count * mem::size_of::<FieldDef>();
+        let fields_ptr = ptr.offset(4);
+        let defaults_ptr = ptr.offset(4).add(fields_size);
```

Comment:
why do you sometimes use `offset` and other times `add` ?

Reply:
Mostly because they take different types, usually `offset` for a constants and `add` for variable-length.